### PR TITLE
[jit] alias_analysis refactor

### PR DESCRIPTION
**Pull request by suo**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17480 [jit] allow "before" and "after" alias annotations&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14214301/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#17511 [jit] alias_analysis refactor**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14231251/)

AliasTracker was doing bookkeeping for three concepts: the points-to graph,
writes, and wildcards.

This PR makes AliasTracker's job clearer: it keeps track of the points-to
graph. Thus it has been renamed MemoryDAG. Write and wildcard information were
pulled back into AliasDb as part of this—I may decide to pull them into their
own little modules since I don't want the alias analysis stuff to get too
bloated.

This refactor is necessary because we want to start tracking information for
aliasing elements that _aren't_ first-class IR Values (e.g. the "stuff" inside
a list). So MemoryDAG can't know too much about Values

Differential Revision: [D14231251](https://our.internmc.facebook.com/intern/diff/D14231251/)

### CODE REVIEW

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 1]
```diff
@@ -1054,5 +1170,34 @@ TORCH_API bool aliasAnalysisHasSpecialCaseFor(Symbol symbol) {
   return handled.count(symbol) || purposefully_not_handled.count(symbol);
 }
 
+// Register `v` as a wildcard value.
+void AliasDb::setWildcard(const Value* v) {
+  wildcards_.insert(v);
+}
+
+void AliasDb::rebuildWriteCache() const {
```
Comment (driazati) [inline torch/csrc/jit/passes/alias_analysis.cpp]: Why is this function `const` if it has no returns/arguments?

File: test/cpp/jit/test_alias_analysis.h [hunk 2]
```diff
@@ -394,6 +395,7 @@ void testAliasAnalysis() {
 
     // Should not be able to move `d` before the wait call
     AliasDb aliasDb(graph);
+    aliasDb.dump();
```
Comment (driazati) [inline test/cpp/jit/test_alias_analysis.h]: remove print from test?

**Review by driazati — APPROVED** [review]

**Comment by zdevito** [comment]
Letting other people review, add me if you think there is something I really need to look at here.

File: test/cpp/jit/test_alias_analysis.h (hunk 2 above)
Comment (suo) [reply test/cpp/jit/test_alias_analysis.h]: oop

File: torch/csrc/jit/passes/alias_analysis.cpp (hunk 1 above)
Comment (suo) [reply torch/csrc/jit/passes/alias_analysis.cpp]: to signify that it does not mutate externally visible state of the object (e.g. is usable from a `const AliasDb`)

**Pull request closed without merging**